Qt backend on BSD fails to compile due to 'GL/gl.h' not found

### Bug Description

This has been reported a long time ago in LibrePCB (https://github.com/LibrePCB/LibrePCB/issues/1593) and I fixed it with [this commit](https://github.com/LibrePCB/slint/commit/73e0bd265a6ea3b35de780e3529c98a75ef4b034) in our Slint fork. Now I'd like to get this fixed upstream.

The problem is, that Slint does not compile with the Qt backend on BSD systems. The original report was about FreeBSD, but I reproduced it with OpenBSD. The procedure is simple:

```
git clone https://github.com/slint-ui/slint.git
cd slint/api/cpp/
mkdir build && cd build
cmake -G Ninja .. -DSLINT_FEATURE_BACKEND_QT=ON -DSLINT_FEATURE_BACKEND_WINIT=OFF
ninja
```

This fails with:

```
warning: i-slint-backend-qt@1.16.0: In file included from qt_window.rs:48:
warning: i-slint-backend-qt@1.16.0: In file included from /usr/local/include/X11/qt6/QtWidgets/QtWidgets:3:
warning: i-slint-backend-qt@1.16.0: In file included from /usr/local/include/X11/qt6/QtWidgets/QtWidgetsDepends:4:
warning: i-slint-backend-qt@1.16.0: In file included from /usr/local/include/X11/qt6/QtGui/QtGui:60:
warning: i-slint-backend-qt@1.16.0: /usr/local/include/X11/qt6/QtGui/qopengl.h:115:13: fatal error: 'GL/gl.h' file not found
warning: i-slint-backend-qt@1.16.0:   115 | #   include <GL/gl.h>
warning: i-slint-backend-qt@1.16.0:       |             ^~~~~~~~~
warning: i-slint-backend-qt@1.16.0: 1 error generated.
error: failed to run custom build command for `i-slint-backend-qt v1.16.0 (/home/slint/internal/backends/qt)`

Caused by:
  process didn't exit successfully: `/home/slint/api/cpp/build/cargo/.._8f183/debug/build/i-slint-backend-qt-122e80ebce879ad1/build-script-build` (exit status: 1)
  --- stdout
  cargo:rustc-check-cfg=cfg(no_qt)
  cargo:rustc-check-cfg=cfg(slint_nightly_test)
  cargo:rerun-if-env-changed=SLINT_NO_QT
  OUT_DIR = Some(/home/slint/api/cpp/build/cargo/.._8f183/x86_64-unknown-openbsd/debug/build/i-slint-backend-qt-597d7b8cfc93f116/out)
  OPT_LEVEL = Some(0)
  TARGET = Some(x86_64-unknown-openbsd)
  CARGO_ENCODED_RUSTFLAGS = Some()
  HOST = Some(x86_64-unknown-openbsd)
  cargo:rerun-if-env-changed=CXX_x86_64-unknown-openbsd
  CXX_x86_64-unknown-openbsd = Some(/usr/bin/c++)
  RUSTC_WRAPPER = None
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some(true)
  cargo:rerun-if-env-changed=CXXFLAGS
  CXXFLAGS = None
  cargo:rerun-if-env-changed=HOST_CXXFLAGS
  HOST_CXXFLAGS = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64_unknown_openbsd
  CXXFLAGS_x86_64_unknown_openbsd = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64-unknown-openbsd
  CXXFLAGS_x86_64-unknown-openbsd = None
  OUT_DIR = Some(/home/slint/api/cpp/build/cargo/.._8f183/x86_64-unknown-openbsd/debug/build/i-slint-backend-qt-597d7b8cfc93f116/out)
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  TARGET = Some(x86_64-unknown-openbsd)
  HOST = Some(x86_64-unknown-openbsd)
  cargo:rerun-if-env-changed=CXXFLAGS
  CXXFLAGS = None
  cargo:rerun-if-env-changed=HOST_CXXFLAGS
  HOST_CXXFLAGS = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64_unknown_openbsd
  CXXFLAGS_x86_64_unknown_openbsd = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64-unknown-openbsd
  CXXFLAGS_x86_64-unknown-openbsd = None
  OUT_DIR = Some(/home/slint/api/cpp/build/cargo/.._8f183/x86_64-unknown-openbsd/debug/build/i-slint-backend-qt-597d7b8cfc93f116/out)
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  TARGET = Some(x86_64-unknown-openbsd)
  HOST = Some(x86_64-unknown-openbsd)
  cargo:rerun-if-env-changed=CXXFLAGS
  CXXFLAGS = None
  cargo:rerun-if-env-changed=HOST_CXXFLAGS
  HOST_CXXFLAGS = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64_unknown_openbsd
  CXXFLAGS_x86_64_unknown_openbsd = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64-unknown-openbsd
  CXXFLAGS_x86_64-unknown-openbsd = None
  OUT_DIR = Some(/home/slint/api/cpp/build/cargo/.._8f183/x86_64-unknown-openbsd/debug/build/i-slint-backend-qt-597d7b8cfc93f116/out)
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  TARGET = Some(x86_64-unknown-openbsd)
  HOST = Some(x86_64-unknown-openbsd)
  cargo:rerun-if-env-changed=CXXFLAGS
  CXXFLAGS = None
  cargo:rerun-if-env-changed=HOST_CXXFLAGS
  HOST_CXXFLAGS = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64_unknown_openbsd
  CXXFLAGS_x86_64_unknown_openbsd = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64-unknown-openbsd
  CXXFLAGS_x86_64-unknown-openbsd = None
  cargo:warning=In file included from qt_window.rs:48:
  cargo:warning=In file included from /usr/local/include/X11/qt6/QtWidgets/QtWidgets:3:
  cargo:warning=In file included from /usr/local/include/X11/qt6/QtWidgets/QtWidgetsDepends:4:
  cargo:warning=In file included from /usr/local/include/X11/qt6/QtGui/QtGui:60:
  cargo:warning=/usr/local/include/X11/qt6/QtGui/qopengl.h:115:13: fatal error: 'GL/gl.h' file not found
  cargo:warning=  115 | #   include <GL/gl.h>
  cargo:warning=      |             ^~~~~~~~~
  cargo:warning=1 error generated.

  --- stderr


  error occurred: ToolExecError: command did not execute successfully (status code exit status: 1): LC_ALL="C" "/usr/bin/c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-gdwarf-2" "-fno-omit-frame-pointer" "-m64" "--target=x86_64-unknown-openbsd" "-I" "/home/slint/internal/backends/qt" "-I" "/usr/local/include/X11/qt6" "-Wall" "-Wextra" "-std=c++17" "-std=c++17" "-o" "/home/slint/api/cpp/build/cargo/.._8f183/x86_64-unknown-openbsd/debug/build/i-slint-backend-qt-597d7b8cfc93f116/out/99a97e52108b2823-cpp_closures.o" "-c" "/home/slint/api/cpp/build/cargo/.._8f183/x86_64-unknown-openbsd/debug/build/i-slint-backend-qt-597d7b8cfc93f116/out/rust_cpp/cpp_closures.cpp"


warning: build failed, waiting for other jobs to finish...
```

The file `GL/gl.h` does exist at `/usr/X11R6/include/GL/gl.h`, butt apparently the include path `/usr/X11R6/include` is not passed to the compiler. My quick fix was to add the corresponding include path, depending on the OS, in `internal/backends/qt/build.rs`:

```rust
    #[cfg(any(target_os = "freebsd", target_os = "dragonfly"))]
    config.include("/usr/local/include");
    #[cfg(target_os = "openbsd")]
    config.include("/usr/X11R6/include");
    #[cfg(target_os = "netbsd")]
    config.include("/usr/pkg/include");
```

(Note that the paths for FreeBSD and OpenBSD have been verified, but not for NetBSD - I think it is actually even wrong).

Anyway, I am not sure if this is the right place for the fix, and the right way to approach the problem. It seems that several more systems require a dedicated include path for the OpenGL header. I figured out this include path is configured in qmake's mkspec files as `QMAKE_INCDIR_OPENGL`. You can see the configured paths like this:

```
cd /usr/lib/qt6/mkspecs/
ag QMAKE_INCDIR_OPENGL

# output:
devices/linux-rasp-pi2-g++/qmake.conf
14:QMAKE_INCDIR_OPENGL_ES2 = $${QMAKE_INCDIR_EGL}

devices/integrity-armv8-drive-cx/qmake.conf
31:QMAKE_INCDIR_OPENGL_ES2 = $${VIBRANTE_INTEGRITY_PDK_TOPDIR}/include

devices/linux-arm-trident-pnx8473-g++/qmake.conf
60:QMAKE_INCDIR_OPENGL_ES2 = $${TRIDENT_SHINER_SDK_INCDIR_EGL_OPENGL_ES2}

devices/linux-beagleboard-g++/qmake.conf
41:QMAKE_INCDIR_OPENGL_ES2 = $${QMAKE_INCDIR_EGL}

devices/freebsd-rasp-pi-clang/qmake.conf
16:QMAKE_INCDIR_OPENGL_ES2 = $${QMAKE_INCDIR_EGL}

devices/linux-rasp-pi3-g++/qmake.conf
27:QMAKE_INCDIR_OPENGL_ES2 = $${QMAKE_INCDIR_EGL}

devices/linux-rasp-pi-g++/qmake.conf
17:QMAKE_INCDIR_OPENGL_ES2 = $${QMAKE_INCDIR_EGL}

devices/integrity-armv8-SA8155P/qmake.conf
38:QMAKE_INCDIR_OPENGL_ES2 = $$QMAKE_INCDIR_EGL

devices/linux-sh4-stmicro-ST7540-g++/qmake.conf
40:QMAKE_INCDIR_OPENGL_ES2 += $$QMAKE_INCDIR_EGL

devices/linux-archos-gen8-g++/qmake.conf
49:QMAKE_INCDIR_OPENGL_ES2 = $${QMAKE_INCDIR_EGL}

devices/integrity-armv8-msm8996au/qmake.conf
54:QMAKE_INCDIR_OPENGL_ES2 = $$(GL_INC_DIR)

devices/linux-arm-hisilicon-hix5hd2-g++/qmake.conf
29:QMAKE_INCDIR_OPENGL_ES2 += $$QMAKE_INCDIR_EGL

devices/linux-mipsel-broadcom-97425-g++/qmake.conf
50:QMAKE_INCDIR_OPENGL_ES2 = $${BRCM_ROCKFORD_PATH}/middleware/v3d/interface/khronos/include

android-clang/qmake.conf
86:QMAKE_INCDIR_OPENGL     =

aix-g++-64/qmake.conf
44:QMAKE_INCDIR_OPENGL     = /usr/X11R6/include

solaris-cc/qmake.conf
48:QMAKE_INCDIR_OPENGL     = /usr/openwin/include

unsupported/linux-host-g++/qmake.conf
89:QMAKE_INCDIR_OPENGL     =
91:QMAKE_INCDIR_OPENGL_ES2 = $$QMAKE_INCDIR_OPENGL

unsupported/vxworks-ppc-dcc/qmake.conf
73:QMAKE_INCDIR_OPENGL     = $$QMAKE_INCDIR_X11

unsupported/vxworks-simpentium-dcc/qmake.conf
73:QMAKE_INCDIR_OPENGL     = $$QMAKE_INCDIR_X11

unsupported/qnx-X11-g++/qmake.conf
22:QMAKE_INCDIR_OPENGL     = /opt/X11R6/include

darwin-g++/qmake.conf
54:QMAKE_INCDIR_OPENGL     = /usr/X11R6/include

hpuxi-g++-64/qmake.conf
52:QMAKE_INCDIR_OPENGL     = /opt/graphics/OpenGL/include /usr/contrib/X11R6/include

aix-g++/qmake.conf
44:QMAKE_INCDIR_OPENGL     = /usr/X11R6/include

cygwin-g++/qmake.conf
44:QMAKE_INCDIR_OPENGL     = /usr/X11R6/include

common/icc-base-unix.conf
111:QMAKE_INCDIR_OPENGL     =

common/uikit/qmake.conf
17:QMAKE_INCDIR_OPENGL_ES2 =
22:QMAKE_INCDIR_OPENGL     =

common/linux.conf
20:QMAKE_INCDIR_OPENGL     =
22:QMAKE_INCDIR_OPENGL_ES2 = $$QMAKE_INCDIR_OPENGL

common/solaris.conf
18:QMAKE_INCDIR_OPENGL     = /usr/X11/include/mesa

common/mac.conf
20:QMAKE_INCDIR_OPENGL     = \

lynxos-g++/qmake.conf
48:QMAKE_INCDIR_OPENGL     = /usr/include/GL

features/unix/opengl.prf
2:    INCLUDEPATH += $$QMAKE_INCDIR_OPENGL_ES2
7:    INCLUDEPATH += $$QMAKE_INCDIR_OPENGL

features/mac/sdk.prf
37:    QMAKE_EXPORT_INCDIR_OPENGL = $$QMAKE_INCDIR_OPENGL
39:    for(val, QMAKE_INCDIR_OPENGL): sysrootified += $${QMAKE_MAC_SDK_PATH}$$val
40:    QMAKE_INCDIR_OPENGL = $$sysrootified

integrity-armv7-imx6/qmake.conf
22:QMAKE_INCDIR_OPENGL_ES2 = $$(GL_INC_DIR)

integrity-armv8-rcar/qmake.conf
29:QMAKE_INCDIR_OPENGL_ES2 = $$QMAKE_INCDIR_EGL

hurd-g++/qmake.conf
20:QMAKE_INCDIR_OPENGL     =
22:QMAKE_INCDIR_OPENGL_ES2 = $$QMAKE_INCDIR_OPENGL

netbsd-g++/qmake.conf
17:QMAKE_INCDIR_OPENGL     = /usr/X11R7/include

openbsd-g++/qmake.conf
17:QMAKE_INCDIR_OPENGL     = /usr/X11R6/include

solaris-cc-64/qmake.conf
65:QMAKE_INCDIR_OPENGL     = /usr/openwin/include
```

So I see different include paths for NetBSD, OpenBSD, Solaris, and other operating systems.

Any suggestions how to fix the compile error? And/or would you accept a PR for just adding the (hardcoded) include paths for FreeBSD, OpenBSD, NetBSD and possibly Solaris to `build.rs`?

### Reproducible Code (if applicable)

```slint

```

### Environment Details

- Slint Version: Current master
- Platform/OS: OpenBSD and others
- Programming Language: C++
- Backend/Renderer: Qt


### Product Impact

_No response_

--- a/internal/backends/qt/build.rs
+++ b/internal/backends/qt/build.rs
@@ -42,6 +42,11 @@ fn main() {
     config.flag_if_supported("/std:c++17");
     // Workaround QTBUG-123153
     config.flag_if_supported("-Wno-template-id-cdtor");
+    // On some systems, the header GL/gl.h (included by some Qt headers) is not
+    // in the compilers default include path, which makes the build fail due to
+    // this header not beeing found. As we don't need OpenGL, we explicitly
+    // disable it with this define. See issue #10989.
+    config.define("QT_NO_OPENGL", None);
     config.include(std::env::var("DEP_QT_INCLUDE_PATH").unwrap()).build("lib.rs");
 
     println!("cargo:rerun-if-changed=lib.rs");

--- a/internal/backends/qt/qt_accessible.rs
+++ b/internal/backends/qt/qt_accessible.rs
@@ -263,6 +263,7 @@ impl SlintAccessibleItemData {
 }
 
 cpp! {{
+    // Note: Do not include <QtWidgets> to avoid inclusion of gl.h (see #10989).
     #include <QtGui/QAccessible>
     #include <QtWidgets/QWidget>
 

--- a/internal/backends/qt/qt_widgets.rs
+++ b/internal/backends/qt/qt_widgets.rs
@@ -147,6 +147,7 @@ impl QImageWrapArray {
 }
 
 cpp! {{
+    // Note: Do not include <QtWidgets> to avoid inclusion of gl.h (see #10989).
     #include <QtWidgets/QApplication>
     #include <QtWidgets/QStyle>
     #include <QtWidgets/QStyleOption>

--- a/internal/backends/qt/qt_window.rs
+++ b/internal/backends/qt/qt_window.rs
@@ -45,26 +45,38 @@ use i_slint_core::renderer::Renderer;
 use std::cell::OnceCell;
 
 cpp! {{
-    #include <QtWidgets/QtWidgets>
-    #include <QtWidgets/QGraphicsScene>
-    #include <QtWidgets/QGraphicsBlurEffect>
-    #include <QtWidgets/QGraphicsPixmapItem>
-    #include <QtGui/QAccessible>
-    #include <QtGui/QPainter>
-    #include <QtGui/QPaintEngine>
-    #include <QtGui/QPainterPath>
-    #include <QtGui/QWindow>
-    #include <QtGui/QResizeEvent>
-    #include <QtGui/QTextLayout>
-    #include <QtGui/QImageReader>
-    #include <QtGui/QCursor>
+    // Note: Do not include <QtWidgets> to avoid inclusion of gl.h (see #10989).
     #include <QtCore/QBasicTimer>
-    #include <QtCore/QTimer>
-    #include <QtCore/QPointer>
     #include <QtCore/QBuffer>
     #include <QtCore/QEvent>
     #include <QtCore/QFileInfo>
+    #include <QtCore/QMutex>
+    #include <QtCore/QPointer>
+    #include <QtCore/QThread>
+    #include <QtCore/QTimer>
+    #include <QtGui/QAccessible>
+    #include <QtGui/QCursor>
+    #include <QtGui/QIconEngine>
+    #include <QtGui/QImageReader>
+    #include <QtGui/QPaintEngine>
+    #include <QtGui/QPainter>
+    #include <QtGui/QPainterPath>
+    #include <QtGui/QResizeEvent>
+    #include <QtGui/QTextLayout>
+    #include <QtGui/QWindow>
+    #include <QtWidgets/QCheckBox>
+    #include <QtWidgets/QComboBox>
+    #include <QtWidgets/QGraphicsBlurEffect>
+    #include <QtWidgets/QGraphicsPixmapItem>
+    #include <QtWidgets/QGraphicsScene>
+    #include <QtWidgets/QGroupBox>
+    #include <QtWidgets/QLineEdit>
+    #include <QtWidgets/QProgressBar>
+    #include <QtWidgets/QPushButton>
+    #include <QtWidgets/QSpinBox>
 
+
+    #include <cmath>
     #include <memory>
 
     void ensure_initialized(bool from_qt_backend);

